### PR TITLE
go 1.21.0

### PR DESCRIPTION
**Pull request by stefanb**
Go 1.21.0 (since all release candidates)

Announcements: 
* rc2: https://groups.google.com/g/golang-announce/c/REcCdQj1taA
* rc3: https://groups.google.com/g/golang-announce/c/5golCy-_yK0
* rc4: https://groups.google.com/g/golang-announce/c/TQs0egfRmQA
* final: https://groups.google.com/g/golang-announce/c/Mk0Jar6hfhI

Release notes: https://tip.golang.org/doc/go1.21
Download: https://go.dev/dl/

## Problems to fix:
* [x] algernon
  * [x] https://github.com/xyproto/algernon/pull/136
  * [x] wait for release >1.15.2: https://github.com/xyproto/algernon/releases
  * [x] bump version in a separate PR: https://github.com/Homebrew/homebrew-core/pull/139176
  * ~~bump via patch: https://github.com/Homebrew/homebrew-core/pull/137320~~
* [ ] [benthos](https://github.com/benthosdev/benthos)
* [ ] brook
  * [ ] upstream https://github.com/txthinking/brook/issues/1245
  * [ ] https://github.com/txthinking/brook/pull/1254
  * [ ] bump version in a separate PR
* [ ] caddy
  * [x] https://github.com/caddyserver/caddy/pull/5644
  * [x] wait for 2.7 caddy version release: https://github.com/caddyserver/caddy/releases
  * [x] version bump in a separate PR #138388 & #138705
* [ ] clair
* [ ] cloudformation-cli
* [x] cloudflared
  * https://github.com/cloudflare/cloudflared/pull/945
  * https://github.com/Homebrew/homebrew-core/pull/136864
* [ ] colima
* [ ] coredns
* [ ] cortex
* [ ] dnscrypt-proxy
  * [x] https://github.com/DNSCrypt/dnscrypt-proxy/issues/2456
  * [x] upstream: https://github.com/DNSCrypt/dnscrypt-proxy/commit/d659a801c2c330c9efe97664114b180b1414e6c6
  * [x] wait for release >2.1.4: https://github.com/DNSCrypt/dnscrypt-proxy/releases
  * [x] bump version in a separate PR #139327
* [ ] [docker-slim](https://github.com/slimtoolkit/slim)
* [x] doggo
  * [x] https://github.com/mr-karan/doggo/pull/72
  * [x] https://github.com/mr-karan/doggo/issues/70
  * https://github.com/mr-karan/doggo/pull/66
  * [x] wait for 0.5.6 release: https://github.com/mr-karan/doggo/releases
  * [x] bump version in a separate PR https://github.com/Homebrew/homebrew-core/pull/137301 & https://github.com/Homebrew/homebrew-core/pull/137303
* [ ] erigon
* [ ] flyctl
* [ ] frpc
* [ ] frps
* [x] [garble](https://github.com/Homebrew/homebrew-core/actions/runs/5370484128/jobs/9742789964?pr=134468#step:4:221): 
  * ~bumped from v0.9.3 to [v0.10.0](https://github.com/burrowers/garble/releases/tag/v0.10.0) via https://github.com/Homebrew/homebrew-core/pull/132902~
  * [x] bump to [v0.10.1](https://github.com/burrowers/garble/releases/tag/v0.10.1) with go 1.21 support by @mvdan 
    * ~https://github.com/Homebrew/homebrew-core/pull/134961~
    * [x] #137087 
  * [X] revision bump in 969dcab436db2c9a7de576463767300a7010e763
  * [x] test adjust in 75b68ca65fc7ef236c9b08671dd736f347a57f0d
* [ ] git-town
* [x] [go-critic](https://github.com/Homebrew/homebrew-core/actions/runs/5370484128/jobs/9742789964?pr=134468#step:4:327): 
  * [x] https://github.com/go-critic/go-critic/issues/1349
    * [x] upstream: https://github.com/quasilyte/go-ruleguard/issues/449
    * upstream: https://github.com/quasilyte/go-ruleguard/pull/451
    * also affected: https://github.com/golangci/golangci-lint/pull/3922
  * [x] https://github.com/go-critic/go-critic/issues/1359
  * [x] bump version in a separate PR https://github.com/Homebrew/homebrew-core/pull/139183
  * [x] re-enable formula
  * [x] revision bump in 46a21881779a4e6e9332c0847ea022f5400b9cd9
* [ ] gollum
* [ ] [gomplate](https://github.com/Homebrew/homebrew-core/actions/runs/5618619997/job/15224884480?pr=134468#step:3:9232)
  * [x] upstream: https://github.com/hairyhenderson/gomplate/pull/1796
  * [ ] bump version in a separate PR
* [ ] hysteria
* [x] [ibazel](https://github.com/Homebrew/homebrew-core/actions/runs/5370484128/jobs/9742789964?pr=134468#step:4:514)
  * [ ] upstream: https://github.com/bazelbuild/bazel-watcher/issues/617
  * [ ] bump version in a separate PR
  * [x] re-enable formula
  * [x] #139337
* [ ] ipfs
* [ ] ipget
* [ ] k3d
* [ ] keploy
* [ ] kubeaudit
* [ ] llama
* [ ] lima
  * [ ] https://github.com/Homebrew/homebrew-core/pull/139393
* [ ] mabel
* [ ] nexttrace
* [ ] ooniprobe
* [ ] portal
* [ ] sdns
* [ ] [sing-box](https://github.com/SagerNet/sing-box)
  * [ ] Wait for 1.4.x release: https://github.com/SagerNet/sing-box/releases
* [ ] syncthing
* [ ] tea
* [ ] terraform-rover
* [ ] thanos
* [ ] tkey-ssh-agent
* [ ] traefik
  * [x] https://github.com/traefik/traefik/issues/10075
* [ ] trezor-bridge
* [ ] twty
* [ ] v2ray
* [ ] xray

Check annotations on https://github.com/Homebrew/homebrew-core/pull/134468/files or action logs.
Those that are failing with `quic-go` errors can likely be fixed by bumping to 0.37.x: eg https://github.com/quic-go/quic-go/releases/tag/v0.37.4
See https://github.com/quic-go/quic-go/releases for latest version.

----

- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [X] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by carlocab** [comment]
Converted this to draft to make sure it doesn't get merged. CI will still run.

File: audit_exceptions/unstable_allowlist.json [hunk 1]
```diff
@@ -3,6 +3,7 @@
   "automysqlbackup": "3.0-rc",
   "aview": "1.3.0rc",
   "ftgl": "2.1.3-rc",
+  "go": "1.21rc",
```
Comment (stefanb) [inline audit_exceptions/unstable_allowlist.json:6]: Temporary, to be reverted once Go 1.21 is released

**Comment by stefanb** [comment]
## [Detected problems](https://github.com/Homebrew/homebrew-core/actions/runs/5370484128/jobs/9742789964?pr=134468) with:
* [ ] [garble](https://github.com/Homebrew/homebrew-core/actions/runs/5370484128/jobs/9742789964?pr=134468#step:4:221): should be bumped from v0.9.3 to [v0.10.0](https://github.com/burrowers/garble/releases/tag/v0.10.0) with go 1.21 support, also `garble was built with "go1.20.5" and is being used with "go1.21rc2"; please rebuild garble with the newer version` https://github.com/Homebrew/homebrew-core/pull/132902
* [ ] [go-critic](https://github.com/Homebrew/homebrew-core/actions/runs/5370484128/jobs/9742789964?pr=134468#step:4:327): https://github.com/go-critic/go-critic/issues/1349 originating in https://github.com/quasilyte/go-ruleguard
* [ ] [ibazel](https://github.com/Homebrew/homebrew-core/actions/runs/5370484128/jobs/9742789964?pr=134468#step:4:514)

File: Formula/garble.rb [hunk 2]
```diff
@@ -17,6 +17,8 @@ class Garble < Formula
     sha256 cellar: :any_skip_relocation, x86_64_linux:   "deb8dd3296bbb92ac22ae64c205a846cfa97631661dbc705be56d6d055355a1d"
   end
 
+  disable! date: "2023-07-15", because: :does_not_build
+
```
Comment (stefanb) [inline Formula/garble.rb:21]: Only temporary, to be removed
```suggestion
```

File: Formula/go-critic.rb [hunk 3]
```diff
@@ -21,6 +21,8 @@ class GoCritic < Formula
     sha256 cellar: :any_skip_relocation, x86_64_linux:   "01aec52aee1170f6fbc2c304d2302884314de73df121c3d7199beb6d80da5aaa"
   end
 
+  disable! date: "2023-07-15", because: :does_not_build
+
```
Comment (stefanb) [inline Formula/go-critic.rb:25]: Only temporary, to be removed
```suggestion
```

File: Formula/ibazel.rb [hunk 4]
```diff
@@ -15,6 +15,8 @@ class Ibazel < Formula
     sha256 cellar: :any_skip_relocation, x86_64_linux:   "a42f541ff46aa784a8f85fafa7d054108d09f242034e1b894febd5a360171e51"
   end
 
+  disable! date: "2023-07-15", because: :does_not_build
+
```
Comment (stefanb) [inline Formula/ibazel.rb:19]: Only temporary, to be removed
```suggestion
```

**Comment by chenrui333** [comment]
for garble, we need to patch the git ref, it has homebrew shim reference issue
for ibazel, https://github.com/bazelbuild/bazel-watcher/issues/617

**Comment by stefanb** [comment]
I tried to write go@1.20 formula twice, but with no luck, that's why I resorted to temporarily disabling these 3 formulae.

**Comment by chenrui333** [comment]
> for garble, we need to patch the git ref, it has homebrew shim reference issue

should be covered by #137087

File: Formula/garble.rb (hunk 2 above)
Comment (stefanb) [reply Formula/garble.rb]: re-enabled in 7bfd23840c445575b4bcf2779bc8d3b18eec535e after cherry picking 6d140c09b77dd33e8cfd2ddfe5354ea20f7582b1 from #137087

File: Formula/go.rb [hunk 5]
```diff
@@ -1,15 +1,16 @@
...
+  sha256 "9de96a2f15196a9c69cb9ab122c9c2aaecca30ec9448caa71543eaeb19ca91ab"
   license "BSD-3-Clause"
   head "https://go.googlesource.com/go.git", branch: "master"
 
   livecheck do
-    url "https://go.dev/dl/"
-    regex(/href=.*?go[._-]?v?(\d+(?:\.\d+)+)[._-]src\.t/i)
+    skip "testing rc2"
+    # url "https://go.dev/dl/"
+    # regex(/href=.*?go[._-]?v?(\d+(?:\.\d+)+)[._-]src\.t/i)
```
Comment (stefanb) [inline Formula/go.rb:13]: To be reverted back to:
```suggestion
    url "https://go.dev/dl/"
    regex(/href=.*?go[._-]?v?(\d+(?:\.\d+)+)[._-]src\.t/i)
```

File: Formula/go.rb [hunk 6]
```diff
@@ -1,15 +1,16 @@
 class Go < Formula
   desc "Open source programming language to build simple/reliable/efficient software"
   homepage "https://go.dev/"
-  url "https://go.dev/dl/go1.20.6.src.tar.gz"
-  mirror "https://fossies.org/linux/misc/go1.20.6.src.tar.gz"
-  sha256 "62ee5bc6fb55b8bae8f705e0cb8df86d6453626b4ecf93279e2867092e0b7f70"
+  url "https://go.dev/dl/go1.21rc3.src.tar.gz"
+  mirror "https://fossies.org/linux/misc/go1.21rc3.src.tar.gz"
+  sha256 "9de96a2f15196a9c69cb9ab122c9c2aaecca30ec9448caa71543eaeb19ca91ab"
```
Comment (stefanb) [inline Formula/go.rb:6]: To be updated with hash from https://go.dev/dl/ once 1.21.0 is released. Check if filenames also have the ".0" suffix!
```suggestion
  url "https://go.dev/dl/go1.21.0.src.tar.gz"
  mirror "https://fossies.org/linux/misc/go1.21.0.src.tar.gz"
  sha256 "...TODO..."
```

File: Formula/go.rb [hunk 7]
```diff
@@ -1,15 +1,18 @@
 class Go < Formula
   desc "Open source programming language to build simple/reliable/efficient software"
   homepage "https://go.dev/"
-  url "https://go.dev/dl/go1.20.7.src.tar.gz"
-  mirror "https://fossies.org/linux/misc/go1.20.7.src.tar.gz"
-  sha256 "2c5ee9c9ec1e733b0dbbc2bdfed3f62306e51d8172bf38f4f4e542b27520f597"
+
```
Comment (stefanb) [inline Formula/go.rb:4]:
```suggestion
```

File: Formula/go.rb [hunk 8]
```diff
@@ -1,15 +1,18 @@
...
+  url "https://go.dev/dl/go1.21rc4.src.tar.gz"
+  mirror "https://fossies.org/linux/misc/go1.21rc4.src.tar.gz"
+  sha256 "2324f20f11112aec3e5d5e428d046861a34e4f99de0ab82a8d914d26b8fb01fd"
+
```
Comment (stefanb) [inline Formula/go.rb:8]:
```suggestion
```

**Comment by chenrui333** [comment]
@stefanb can you either rebase to the latest master or create a new PR? Thanks!

File: Formula/go-critic.rb (hunk 3 above)
Comment (chenrui333) [reply Formula/go-critic.rb:25]: might be good to add a note in the code to reference the open issue, `https://github.com/go-critic/go-critic/issues/1349`

**Comment by stefanb** [comment]
@chenrui333 rebased.

**Comment by jmhodges** [comment]
The ibazel blocker seem to be corrected by a version bump of the `rules_go` dep in ibazel's formula test to v0.41. [Longer explanation](https://github.com/bazelbuild/bazel-watcher/issues/617#issuecomment-1670228126).

**Comment by ivan-penchev** [comment]
Go 1.21 was released yesterday. Do you need something specific so this formula update could be merged? @stefanb @chenrui333

**Comment by mrkagelui** [comment]
Can't wait to have this, thank you!

**Review by chenrui333 — APPROVED** [review]

**Comment by chenrui333** [comment]
should be good to go (and we followup with go@1.20 and other stuff)

**Comment by mrkagelui** [comment]
is there anything we can help to fix any issue?

**Comment by stefanb** [comment]
> is there anything we can help to fix any issue?

@mrkagelui and others can help updating dependant projects and formulae to be compatible with Go 1.21. See the list of projects (with issues and PRs) in this PR description and in annotations at the bottom of [changed files in this PR](https://github.com/Homebrew/homebrew-core/pull/134468/files).

**Comment by mrkagelui** [comment]
hmm... I've no experience in maintaining this, let me see what I can do. right now I can't figure out why some tests are failing after hours

**Comment by stefanb** [comment]
Builds are timing out because Go has many dependants.

Some labels could probably help to get the build complete and see all the failing formulae:
* [CI-long-timeout](https://github.com/Homebrew/homebrew-core/pulls?q=label%3ACI-long-timeout)
* [long build](https://github.com/Homebrew/homebrew-core/issues?q=label%3A%22long+build%22)

**Comment by uudashr** [comment]
What are we waiting for? is there any action items left?

**Comment by islishude** [comment]
Run `brew uninstall go` and get it from https://go.dev/dl/

**Comment by jackmcguire1** [comment]
bump

**Comment by stefanb** [comment]
Fully agree :shipit:

**Comment by jedisct1** [comment]
`dnscrypt-proxy` 2.1.5 released. Should be compatible with Go 1.21.

**Comment by stefanb** [comment]
@jedisct1 tnx for the notification. To not block this PR i created a separate PR:
* #139327

**Comment by pseidemann** [comment]
naive question about the process in general: wouldn't it be possible to always create a new `go@1.x` formula, which could be released basically on the same day as the official release, without changing any other formulae? the main `go` formula could be then changed some time in the future.

**Comment by sonu27** [comment]
@pseidemann I agree somewhat, it is annoying having to always wait for the new release. Wish this would change.

**Comment by Neo2308** [comment]
@fxcoudert/@p-linnane  can you review this PR? (Sorry for the tag)

**Comment by AkihiroSuda** [comment]
Released Lima v0.17.1 to support Go 1.21
- https://github.com/Homebrew/homebrew-core/pull/139393

**Comment by chenrui333** [comment]
this is the list (https://github.com/Homebrew/homebrew-core/actions/runs/5832665576/job/15819772214) we can followup, let's ship it for now.

**Comment by chenrui333** [comment]
Thanks @stefanb and everyone, that was a good run. 🚀

**Comment by chenrui333** [comment]
my checklist on ventura arm machine

- brook, quic-go-and-Go-versions, https://github.com/txthinking/brook/pull/1254
- caddy, quic-go@v0.37.3, cannot use h.allow0RTT, https://github.com/caddyserver/caddy/pull/5719
- cloudflared, quic-go-and-Go-versions, https://github.com/cloudflare/cloudflared/issues/1043
- colima, qemu issue
- coredns, quic-go@v0.37.3, cannot use h.allow0RTT, https://github.com/coredns/coredns/pull/6252
- dnscrypt-proxy, good
- doggo, quic-go@v0.37.3, cannot use h.allow0RTT, https://github.com/mr-karan/doggo/pull/76
- erigon, https://github.com/ledgerwatch/erigon/issues/7984
```
/Users/rui/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/anacrolix/go-libutp@v1.3.0/callbacks.go:34:10: cannot define new methods on non-local type *C.utp_callback_arguments
/Users/rui/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/anacrolix/go-libutp@v1.3.0/callbacks.go:38:10: cannot define new methods on non-local type *C.utp_callback_arguments
/Users/rui/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/anacrolix/go-libutp@v1.3.0/utp.go:29:12: cannot define new methods on non-local type *C.utp_context
/Users/rui/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/anacrolix/go-libutp@v1.3.0/utp.go:40:12: cannot define new methods on non-local type *C.utp_context
```
- flyctl, https://github.com/superfly/flyctl/issues/2688
- frpc, quic-go-and-Go-versions, https://github.com/fatedier/frp/issues/3572
- frps, quic-go-and-Go-versions, https://github.com/fatedier/frp/issues/3572
- gollum, https://github.com/trivago/gollum/issues/265
- hysteria, quic-go-and-Go-versions, https://github.com/apernet/hysteria/issues/647
- ibazel, fixed
- ipfs, quic-go-and-Go-versions, https://github.com/ipfs/kubo/issues/10064
- ipget, quic-go-and-Go-versions, https://github.com/ipfs/ipget/issues/141
- k3d, go4.org/unsafe/assume-no-moving-gc, https://github.com/k3d-io/k3d/issues/1332
- lima, qemu
- mabel, https://github.com/smmr-software/mabel/issues/15
- ooniprobe, quic-go-and-Go-versions
- sdns, good
- sing-box, quic-go-and-Go-versions, https://github.com/SagerNet/sing-box/issues/774
- syncthing, quic-go-and-Go-versions, https://github.com/syncthing/syncthing/issues/9034
- terracognita, works
- terraform-rover, node issue
- tkey-ssh-agent, build failure (`usb_darwin`)
- traefik, quic-go-and-Go-versions, https://github.com/traefik/traefik/issues/10075
- trezor-bridge, `libusb`
- v2ray, quic-go-and-Go-versions, https://github.com/v2fly/v2ray-core/issues/2644
- xray, build issue

**Pull request merged**